lib/routes/sign.js

```diff
@@ -49,104 +45,86 @@ module.exports = function (log, isA, error, signer, db, domain, metricsContext,
-                  return reply(error.missingRequestParameter('p'))
-                }
-                if (!publicKey.q) {
-                  return reply(error.missingRequestParameter('q'))
-                }
-                if (!publicKey.g) {
-                  return reply(error.missingRequestParameter('g'))
-                }
-              }
+        if (publicKey.algorithm === 'RS') {
+          if (!publicKey.n) {
+            return reply(error.missingRequestParameter('n'))
+          }
+          if (!publicKey.e) {
+            return reply(error.missingRequestParameter('e'))
+          }
+        }
+        else { // DS
+          if (!publicKey.y) {
+            return reply(error.missingRequestParameter('y'))
+          }
+          if (!publicKey.p) {
+            return reply(error.missingRequestParameter('p'))
+          }
+          if (!publicKey.q) {
+            return reply(error.missingRequestParameter('q'))
+          }
+          if (!publicKey.g) {
+            return reply(error.missingRequestParameter('g'))
+          }
+        }
 
-              if (!sessionToken.locale) {
-                if (request.app.acceptLanguage) {
-                  // Log details to sanity-check locale backfilling.
-                  log.info({
-                    op: 'signer.updateLocale',
-                    locale: request.app.acceptLanguage
-                  })
-                  db.updateLocale(sessionToken.uid, request.app.acceptLanguage)
-                  // meh on the result
-                } else {
-                  // We're seeing a surprising number of accounts that don't get
-                  // a proper locale.  Log details to help debug this.
-                  log.info({
-                    op: 'signer.emptyLocale',
-                    email: sessionToken.email,
-                    locale: request.app.acceptLanguage,
-                    agent: request.headers['user-agent']
-                  })
-                }
-              }
-              uid = sessionToken.uid.toString('hex')
+        if (!sessionToken.locale) {
+          if (request.app.acceptLanguage) {
+            // Log details to sanity-check locale backfilling.
+            log.info({
+              op: 'signer.updateLocale',
+              locale: request.app.acceptLanguage
+            })
+            db.updateLocale(sessionToken.uid, request.app.acceptLanguage)
+            // meh on the result
+          } else {
+            // We're seeing a surprising number of accounts that don't get
+            // a proper locale.  Log details to help debug this.
+            log.info({
+              op: 'signer.emptyLocale',
+              email: sessionToken.email,
+              locale: request.app.acceptLanguage,
+              agent: request.headers['user-agent']
+            })
+          }
+        }
+        var uid = sessionToken.uid.toString('hex')
+        var deviceId = sessionToken.deviceId ? sessionToken.deviceId.toString('hex') : null
```

Comment:
It is not obvious to me why a device might not have an id? Would it make sense to log details?